### PR TITLE
Make the number of axes for batch dimensions configurable 

### LINKED ISSUE #2019: Make the number of axes for batch dimensions configurable in Linear
L.Linear accepts multiple channel dimensions. It is useful in computer vision applications where the output of the final convolution layer can be directly fed to a linear link. This feature can be seen as a flexible number of channel axes.

On the other hand, the number of batch axes is currently fixed to one, which would be cumbersome in some use cases, especially when we have to deal with a sequence of input arrays. In this case, the first two axes may be used to represent the mini-batch and the sequence. We currently need to reshape it before and after feeding it to the linear link, which has overheads both for coding and runtime.

It can be resolved by introducing an option to Linear (and also to the F.linear function) that indicates the number of batch axes.

The same discussions would also be applied to convolution links/functions.

### PR DESCRIPTION
**Pull request by mitmul**
Fix #2019 

This PR enables to specify the number of batch axes when you use L.Linear or F.huber_loss.
For example, with this PR, users can input (4, 2, 3)-shaped input to `L.Linear(3, 2)` and get (4, 2, 2)-shaped output.

- [x] Write tests for Linear link/function
- [x] F.huber_loss + tests

### CODE REVIEW

**Comment by mitmul** [comment]
I think adding n_batch_axes option can be realized without breaking the backward compatibility after v2. I'll rewrite this PR to add n_batch_axes argument with the default value of 1 as the final optional keyword argment.

File: chainer/functions/loss/huber_loss.py [hunk 1]
```diff
@@ -7,21 +10,31 @@
 
 class HuberLoss(function.Function):
 
-    def __init__(self, delta, reduce='sum_along_second_axis'):
+    def __init__(self, delta, n_batch_axes=1, reduce='sum_each_data'):
```
Comment (unnonouno) [inline chainer/functions/loss/huber_loss.py]: Change order of argument. It breaks compatibility.

File: chainer/functions/loss/huber_loss.py [hunk 2]
```diff
@@ -43,13 +57,14 @@ def backward(self, inputs, gy):
...
-            gy_ = gy_.reshape(gy[0].shape + (1,) * (self.diff.ndim - 1))
+        if self.reduce == 'sum_each_data':
+            gy_ = gy_.reshape(
+                gy_.shape + (1,) * (self.diff.ndim - self._n_batch_axes))
         gx = gy_ * gx
         return gx, -gx
 
 
-def huber_loss(x, t, delta, reduce='sum_along_second_axis'):
+def huber_loss(x, t, delta, n_batch_axes=1, reduce='sum_each_data'):
```
Comment (unnonouno) [inline chainer/functions/loss/huber_loss.py]: ditto

File: chainer/functions/connection/linear.py [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
+        return x.reshape(x.shape[:n_batch_axes] + (-1,))
+    else:
+        raise ValueError('n_batch_axes should be less than x.ndim and greater '
+                         'than 0 but {} was given.'.format(n_batch_axes))
 
 
 class LinearFunction(function.Function):
 
+    def __init__(self, n_batch_axes=1):
+        self._n_batch_axes = n_batch_axes
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: Is it safe when n_batch_axes has negative value?

File: chainer/functions/connection/linear.py [hunk 4]
```diff
@@ -1,15 +1,25 @@
...
-def _as_mat(x):
-    if x.ndim == 2:
+def _as_mat(x, n_batch_axes):
+    if n_batch_axes == 1:
         return x
-    return x.reshape(len(x), -1)
+    elif 1 < n_batch_axes < x.ndim:
+        return x.reshape(x.shape[:n_batch_axes] + (-1,))
+    else:
+        raise ValueError('n_batch_axes should be less than x.ndim and greater '
```
Comment (unnonouno) [inline chainer/functions/connection/linear.py]: Write test case for this case.

File: chainer/functions/loss/huber_loss.py [hunk 5]
```diff
@@ -7,21 +10,31 @@
 
 class HuberLoss(function.Function):
 
-    def __init__(self, delta, reduce='sum_along_second_axis'):
+    def __init__(self, delta, n_batch_axes=1, reduce='sum_each_data'):
         self.delta = delta
+        self._n_batch_axes = n_batch_axes
+
+        if reduce == 'sum_along_second_axis':
+            warnings.warn("sum_along_second_axis for the 'reduce' argment is"
```
Comment (unnonouno) [inline chainer/functions/loss/huber_loss.py]: Write test for this case. You can check warning using `warnings` package.

File: chainer/functions/connection/linear.py (hunk 4 above)
Comment (mitmul) [inline chainer/functions/connection/linear.py]: Thanks, I added test cases.

File: chainer/functions/loss/huber_loss.py (hunk 5 above)
Comment (mitmul) [inline chainer/functions/loss/huber_loss.py]: Thanks, I added `TestHuberLossInvalidReductionOption`

**Comment by mitmul** [comment]
@unnonouno Thank you for your review. I fixed what you pointed out.

**Comment by unnonouno** [comment]
It is conflicted. Can you split this PR into two PRs? And please resolve conflicts.

**Comment by mitmul** [comment]
The introduction of FunctionNode dramatically changed the code around Linear function, then many conflicts happened in this PR. It's hard to fix them from here, so I will send different PRs for Linear and HuberLoss separately.

**Comment by niboshi** [comment]
@mitmul Should we close this PR?

**Comment by mitmul** [comment]
Yes, it's better to make another PR to add this feature. I close this.

**Pull request closed without merging**